coreneuron/mpi/spdlog_mpi_logger.h

```diff
@@ -0,0 +1,92 @@
+/* Example
+#include <spdlog_mpi_logger.h>
+
+int main(int argc, char *argv[]) {
+    MPI_Init(&argc, &argv);
+    auto logger = std::make_shared<spdlog::mpi_logger>("my_mpi_logger");
+    logger->info("ceci est un test");
+    MPI_Finalize();
+    return 0;
+}
+*/
+
+#pragma once
+
+#include <iostream>
+#include <spdlog/logger.h>
+#include <mpi.h>
+
+namespace spdlog {
+
+class mpi_logger: public logger
+{
+public:
+    // Empty logger
+    explicit mpi_logger(std::string logger_name, int tag_ = 4444, int server_rank_ = 0, MPI_Comm comm_ = MPI_COMM_WORLD)
+        : logger(std::move(logger_name))
+        , tag(tag_), server_rank(server_rank_), comm(comm_)
+    {
+        MPI_Comm_rank(comm, &mpi_rank);
+        if (mpi_rank == server_rank) {
+            server_thr = std::thread([&](){
+                int i = 4;
+                while(i--) {
+                    MPI_Status status;
+                    MPI_Probe(MPI_ANY_SOURCE, tag, comm, &status);
+                    int msg_size = 0;
+                    MPI_Get_count(&status, MPI_BYTE, &msg_size);
+                    auto buf = std::make_unique<char[]>(msg_size);
+                    MPI_Recv(buf.get(), msg_size, MPI_PACKED, MPI_ANY_SOURCE, tag, comm, MPI_STATUS_IGNORE);
+
+                    int position = 0;
+                    int level = 0;
+                    MPI_Unpack(buf.get(), msg_size, &position, &level, 1, MPI_INT, comm);
+                    unsigned long payload_size = 0;
+                    MPI_Unpack(buf.get(), msg_size, &position, &payload_size, 1, MPI_UNSIGNED_LONG, comm);
+                    std::string payload(payload_size, ' ');
+                    MPI_Unpack(buf.get(), msg_size, &position, &payload[0], payload_size, MPI_BYTE, comm);
+                    std::cout << "[" << level << "] " << payload << std::endl;
```

Comment:
* MPI_Probe, MPI_Recv are blocking calls
* It's true that these are running under separate thread and hence argument could be they won't block the main thread
* then, note that this requires enabling thread safety in MPI library (which comes with the cost)
* I wouldn't claim every MPI library is fully thread safe (e.g. for `MPI_THREAD_MULTIPLE` and related performance penalty)
* (I understand that collecting logs at `server_rank` is one `policy` but this won't be scalable.)

Reply:
Yes, I made them blocking because this thread has nothing else to do.
hm, I will look for MPI thread safety, but it seems me unrelated there.
why, is it not scallable? do you have an idea on an other policy?